File: android/app/src/main/java/io/openorcha/mobile/data/OrchaApiClient.kt
package io.openorcha.mobile.data

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.plugins.HttpTimeoutConfig
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.client.request.patch
import io.ktor.client.request.post
import io.ktor.client.request.prepareGet
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsChannel
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.encodeURLParameter
import io.ktor.serialization.kotlinx.json.json
import io.ktor.utils.io.readUTF8Line
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json

class OrchaApiClient {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        explicitNulls = false
    }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(json) }
        install(HttpTimeout) {
            requestTimeoutMillis = 10_000
            connectTimeoutMillis = 3_000
            socketTimeoutMillis = 10_000
        }
    }

    suspend fun listContainers(baseUrl: String): ContainersResponse = withTimeout(6_000) {
        client.get("${baseUrl.endpoint()}/api/containers").body()
    }

    suspend fun getSnapshot(
        baseUrl: String,
        containerId: String,
        taskLimit: Int? = null,
        requestLimit: Int? = null,
    ): ContainerSnapshot = withTimeout(10_000) {
        val window = listOfNotNull(
            taskLimit?.let { "task_limit=$it" },
            requestLimit?.let { "request_limit=$it" },
        ).joinToString("&").let { if (it.isEmpty()) "" else "?$it" }
        client.get("${baseUrl.endpoint()}/api/containers/$containerId$window").body()
    }

    /** Keyset-paged thread fetch: newest page first; older pages via before/before_id. */
    suspend fun getTaskMessages(
        baseUrl: String,
        taskId: String,
        limit: Int = THREAD_PAGE,
        before: String? = null,
        beforeId: String? = null,
    ): TaskMessagesResponse = withTimeout(8_000) {
        val cursor = listOfNotNull(
            before?.let { "before=${it.encodeURLParameter()}" },
            beforeId?.let { "before_id=${it.encodeURLParameter()}" },
        ).joinToString("&").let { if (it.isEmpty()) "" else "&$it" }
        client.get("${baseUrl.endpoint()}/api/tasks/$taskId/messages?limit=$limit$cursor").body()
    }

    suspend fun postTaskMessage(baseUrl: String, taskId: String, actorId: String, body: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/tasks/$taskId/messages", TaskMessageBody(actorId, body))

    suspend fun cancelTask(baseUrl: String, taskId: String, actorId: String, reason: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/tasks/$taskId/cancel", TaskCancelBody(actorId, reason.blankToNull()))

    suspend fun verifyTask(baseUrl: String, taskId: String, actorId: String, approve: Boolean, feedback: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/tasks/$taskId/verify", TaskVerifyBody(approve, feedback.blankToNull(), actorId))

    suspend fun decidePlan(
        baseUrl: String,
        taskId: String,
        actorId: String,
        approve: Boolean,
        reason: String?,
        targetAgentId: String?,
    ): GenericIdResponse =
        postJson(
            "${baseUrl.endpoint()}/api/decisions",
            DecisionBody(
                subjectType = "plan_approval",
                subjectId = taskId,
                decision = if (approve) "approve" else "reject",
                reason = reason.blankToNull(),
                actorAgentId = actorId,
                targetAgentId = targetAgentId,
            ),
        )

    suspend fun getTaskRuns(baseUrl: String, taskId: String): RunsResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/tasks/$taskId/runs?limit=$RUNS_PAGE").body()
    }

    // flow 09 lazy sections
    suspend fun getPersona(baseUrl: String, agentId: String): PersonaResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/persona").body()
    }

    suspend fun getDigest(baseUrl: String, agentId: String): DigestResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/digest").body()
    }

    suspend fun getInbox(baseUrl: String, agentId: String): InboxResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/inbox").body()
    }

    suspend fun getOutbox(baseUrl: String, agentId: String): OutboxResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/outbox").body()
    }

    suspend fun getResidentRuns(baseUrl: String, agentId: String): RunsResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/resident-runs").body()
    }

    suspend fun updateAgent(baseUrl: String, agentId: String, actorId: String, alias: String?, role: String?): GenericIdResponse =
        patchJson("${baseUrl.endpoint()}/api/agents/$agentId", AgentUpdateBody(actorId, alias.blankToNull(), role.blankToNull()))

    // flow 05: implications preview before the destructive close confirm
    suspend fun getCloseImplications(baseUrl: String, taskId: String): CloseImplicationsResponse = withTimeout(5_000) {
        client.get("${baseUrl.endpoint()}/api/tasks/$taskId/close-implications").body()
    }

    suspend fun getAgentRuns(baseUrl: String, agentId: String): RunsResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/runs?limit=$RUNS_PAGE").body()
    }

    /**
     * FINISHED runs only: the server closes the stream immediately, so a buffered read is
     * safe. A RUNNING run must go through [streamRun] — this call's timeouts would kill it.
     */
    suspend fun getRunStreamText(baseUrl: String, agentId: String, runId: String): String = withTimeout(20_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/runs/$runId/stream").bodyAsText()
    }

    /**
     * Live run-log stream (issue 3): same endpoint the web's EventSource consumes, read
     * incrementally. The per-request infinite timeout is load-bearing — the client's
     * global 10s request cap would otherwise kill a stream that never ends; no coroutine
     * withTimeout may wrap this call (that was the old getRunStreamText bug).
     */
    fun streamRun(baseUrl: String, agentId: String, runId: String): Flow<RunStreamEvent> = flow {
        client.prepareGet("${baseUrl.endpoint()}/api/agents/$agentId/runs/$runId/stream") {
            timeout {
                requestTimeoutMillis = HttpTimeoutConfig.INFINITE_TIMEOUT_MS
                socketTimeoutMillis = HttpTimeoutConfig.INFINITE_TIMEOUT_MS
            }
        }.execute { response ->
            val channel = response.bodyAsChannel()
            while (true) {
                val raw = channel.readUTF8Line() ?: break
                RunStream.parse(raw)?.let { emit(it) }
            }
        }
    }

    suspend fun stopRun(baseUrl: String, runId: String, actorId: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/runs/$runId/stop", WorkerRunStopBody(actorId))

    suspend fun respondRequest(baseUrl: String, requestId: String, actorId: String, response: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/respond", RequestRespondBody(actorId, response))

    suspend fun closeRequest(baseUrl: String, requestId: String, actorId: String, reason: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/close", RequestActorBody(actorId, reason.blankToNull()))

    suspend fun nudgeRequest(baseUrl: String, requestId: String, actorId: String, note: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/nudge", NudgeBody(actorId, note.blankToNull()))

    suspend fun escalateRequest(baseUrl: String, requestId: String, actorId: String, reason: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/escalate", RequestActorBody(actorId, reason.blankToNull()))

    suspend fun acceptTaskRequest(baseUrl: String, requestId: String, actorId: String, note: String?): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/accept-task", TaskRequestAcceptBody(actorId, note.blankToNull()))

    suspend fun rejectTaskRequest(baseUrl: String, requestId: String, actorId: String, reason: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/requests/$requestId/reject-task", TaskRequestRejectBody(actorId, reason))

    suspend fun convertRequest(
        baseUrl: String,
        requestId: String,
        actorId: String,
        title: String,
        definitionOfDone: String,
        assigneeAlias: String?,
        priority: Int = 100,
    ): GenericIdResponse =
        postJson(
            "${baseUrl.endpoint()}/api/requests/$requestId/convert-to-task",
            RequestConvertBody(actorId, title, definitionOfDone, priority, assigneeAlias.blankToNull()),
        )

    suspend fun listModels(baseUrl: String): ModelsResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/models").body()
    }

    suspend fun updateAgentModel(baseUrl: String, agentId: String, model: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/agents/$agentId/model", AgentModelBody(model))

    suspend fun updateAutoWake(baseUrl: String, agentId: String, actorId: String, intervalSecs: Int?): GenericIdResponse =
        patchJson("${baseUrl.endpoint()}/api/agents/$agentId/auto-wake", AutoWakeBody(actorId, intervalSecs))

    suspend fun retireAgent(baseUrl: String, agentId: String, actorId: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/agents/$agentId/retire", AgentRetireBody(actorId))

    suspend fun getConversation(baseUrl: String, agentId: String): ConversationResponse = withTimeout(8_000) {
        client.get("${baseUrl.endpoint()}/api/agents/$agentId/conversation?limit=80").body()
    }

    /** Delta refresh (web conversation.js:586): only turns after the last seen seq. */
    suspend fun getConversationTurns(baseUrl: String, conversationId: String, afterSeq: Int, limit: Int = 50): TurnsResponse =
        withTimeout(8_000) {
            client.get("${baseUrl.endpoint()}/api/conversations/$conversationId/turns?after_seq=$afterSeq&limit=$limit").body()
        }

    suspend fun startConversation(baseUrl: String, agentId: String, actorId: String): ConversationResponse =
        postJson("${baseUrl.endpoint()}/api/agents/$agentId/conversations", ConversationStartBody(actorId))

    suspend fun sendConversationTurn(baseUrl: String, conversationId: String, actorId: String, content: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/conversations/$conversationId/turns", TurnAppendBody("human", actorId, content))

    suspend fun endConversation(baseUrl: String, conversationId: String, actorId: String): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/conversations/$conversationId/end", ConversationActorBody(actorId))

    suspend fun createTask(
        baseUrl: String,
        containerId: String,
        title: String,
        description: String?,
        definitionOfDone: String,
        actorId: String,
        assigneeAlias: String?,
        priority: Int,
        dependsOn: List<String>,
        notReady: Boolean,
    ): GenericIdResponse =
        postJson(
            "${baseUrl.endpoint()}/api/containers/$containerId/tasks",
            TaskCreateBody(
                title = title,
                description = description.blankToNull(),
                definitionOfDone = definitionOfDone,
                priority = priority,
                createdByAgentId = actorId,
                assigneeAlias = assigneeAlias.blankToNull(),
                dependsOn = dependsOn,
                notReady = notReady,
            ),
        )

    suspend fun assignTask(baseUrl: String, taskId: String, actorId: String, agentId: String, reassign: Boolean): GenericIdResponse =
        postJson("${baseUrl.endpoint()}/api/tasks/$taskId/assign", AssignTaskBody(actorId, agentId, reassign))

    private suspend inline fun <reified T : Any, reified R> postJson(url: String, payload: T): R = withTimeout(10_000) {
        val response: HttpResponse = client.post(url) {
            contentType(ContentType.Application.Json)
            setBody(payload)
        }
        response.body()
    }

    private suspend inline fun <reified T : Any, reified R> patchJson(url: String, payload: T): R = withTimeout(10_000) {
        val response: HttpResponse = client.patch(url) {
            contentType(ContentType.Application.Json)
            setBody(payload)
        }
        response.body()
    }

    private fun String.endpoint(): String = OrchaServerAddress.normalize(this)
    private fun String?.blankToNull(): String? = this?.trim()?.takeIf { it.isNotEmpty() }

    companion object {
        /** Task-thread keyset page size (issue 4; the thread fetch was the one unbounded call). */
        const val THREAD_PAGE = 20

        /** Explicit runs-list window (documents the server default). */
        const val RUNS_PAGE = 20
    }
}

File: ios/Orcha/Data/OrchaApiClient.swift
import Foundation

/// One parsed frame of the worker-run SSE stream (`main.py:5671-5673`):
/// `data: {"seq","line"}` progress lines and a terminal `data: {"seq","done","status"}`.
enum RunStreamEvent {
    case line(seq: Int, text: String)
    case done(seq: Int, status: String)
}

/// Thin async URLSession client over the Orcha REST surface — the same endpoints
/// the Android client is proven against. All calls throw on non-2xx.
struct OrchaApiClient {
    private let session: URLSession
    /// A separate, un-timed-out session for the run-log SSE stream: a running run never
    /// closes, so the 20s `timeoutIntervalForResource` on `session` would kill it (Issue 3).
    private let streamSession: URLSession
    private let decoder = JSONDecoder()
    /// Explicit credential override — used by the add-flow probe, where the token the
    /// user just pasted isn't persisted (and so isn't in `BearerTokens`) yet. Nil for
    /// the app-wide client: it resolves per request from the registry.
    private let bearerOverride: String?

    init(bearerToken: String? = nil) {
        bearerOverride = bearerToken
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 20
        session = URLSession(configuration: config)

        let streamConfig = URLSessionConfiguration.ephemeral
        // No caps: 1s heartbeat comments keep the socket live; the server itself caps each
        // open at 30 min (`stream_timeout`), after which the collector reopens.
        streamConfig.timeoutIntervalForRequest = .infinity
        streamConfig.timeoutIntervalForResource = .infinity
        streamSession = URLSession(configuration: streamConfig)
    }

    // MARK: reads

    func listContainers(_ base: String) async throws -> ContainersResponse {
        try await get(base, "/api/containers")
    }

    func snapshot(_ base: String, _ cid: String, taskLimit: Int? = nil, requestLimit: Int? = nil) async throws -> ContainerSnapshot {
        try await get(base, "/api/containers/\(cid)" + query([
            "task_limit": taskLimit.map(String.init),
            "request_limit": requestLimit.map(String.init),
        ]))
    }

    /// Issue 4: with `limit` the endpoint returns the NEWEST `limit` messages (ASC within the
    /// page) plus a `(next_before, next_before_id)` keyset cursor to load earlier pages.
    func taskMessages(_ base: String, _ tid: String, limit: Int? = nil, before: String? = nil, beforeId: String? = nil) async throws -> TaskMessagesResponse {
        try await get(base, "/api/tasks/\(tid)/messages" + query([
            "limit": limit.map(String.init),
            "before": before,
            "before_id": beforeId,
        ]))
    }

    func taskRuns(_ base: String, _ tid: String, limit: Int = 20) async throws -> RunsResponse {
        try await get(base, "/api/tasks/\(tid)/runs" + query(["limit": String(limit)]))
    }

    func agentRuns(_ base: String, _ aid: String, limit: Int = 20) async throws -> RunsResponse {
        try await get(base, "/api/agents/\(aid)/runs" + query(["limit": String(limit)]))
    }

    func residentRuns(_ base: String, _ aid: String) async throws -> RunsResponse {
        try await get(base, "/api/agents/\(aid)/resident-runs")
    }

    func persona(_ base: String, _ aid: String) async throws -> PersonaResponse {
        try await get(base, "/api/agents/\(aid)/persona")
    }

    func digest(_ base: String, _ aid: String) async throws -> DigestResponse {
        try await get(base, "/api/agents/\(aid)/digest")
    }

    func inbox(_ base: String, _ aid: String) async throws -> InboxResponse {
        try await get(base, "/api/agents/\(aid)/inbox")
    }

    func outbox(_ base: String, _ aid: String) async throws -> OutboxResponse {
        try await get(base, "/api/agents/\(aid)/outbox")
    }

    func models(_ base: String) async throws -> ModelsResponse {
        try await get(base, "/api/models")
    }

    /// Collab v1 — who is the acting human on this project, per the trusted proxy
    /// identity (device-token requests carry it too; identity_routes.py).
    func me(_ base: String, _ cid: String) async throws -> MeResponse {
        try await get(base, "/api/me" + query(["cid": cid]))
    }

    /// Collab v1 — the project's human roster (`restricted:true` = own row only).
    func members(_ base: String, _ cid: String) async throws -> MembersResponse {
        try await get(base, "/api/containers/\(cid)/members")
    }

    /// mig 040 — the signed-in user's cosmetic prefs bag, or nil (the deliberate
    /// self-host / trust-off / unmapped "stay local-only" signal — a 200, not an error).
    func getPrefs(_ base: String) async throws -> [String: Any]? {
        let (data, _) = try await raw(base, "/api/prefs")
        let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        return obj?["prefs"] as? [String: Any]
    }

    /// mig 040 — whole-bag replace of the signed-in user's cosmetic prefs.
    func putPrefs(_ base: String, _ prefs: [String: Any]) async throws {
        _ = try await send(base, "/api/prefs", method: "PUT", ["prefs": prefs])
    }

    /// The GitHub App installation's repo list for the Connect-repo sheet
    /// (portal `home-github.js` parity). `available:false` rides a 200 — a
    /// graceful off state, never thrown. `cid` scopes the listing to the selected
    /// project so the server can fall back to that project's saved PAT (the React
    /// client sends it too); without it a PAT-only local project lists nothing.
    func githubRepos(_ base: String, cid: String? = nil) async throws -> GithubReposResponse {
        try await get(base, githubReposPath(cid: cid))
    }

    /// `/api/github/repos` (+ `?cid=` when a project is selected) — split out so the
    /// query contract is unit-testable without a network.
    func githubReposPath(cid: String?) -> String {
        "/api/github/repos" + query(["cid": cid])
    }

    func conversation(_ base: String, _ aid: String, limit: Int? = nil) async throws -> ConversationResponse {
        try await get(base, "/api/agents/\(aid)/conversation" + query(["limit": limit.map(String.init)]))
    }

    /// Issue 4: delta-append the turns created after `afterSeq` (oldest→newest), instead of
    /// full-replacing the whole conversation on every refresh.
    func conversationTurns(_ base: String, _ convId: String, afterSeq: Int, limit: Int = 50) async throws -> TurnsResponse {
        try await get(base, "/api/conversations/\(convId)/turns" + query([
            "after_seq": String(afterSeq), "limit": String(limit),
        ]))
    }

    /// One-shot read of a FINISHED run's log (the server closes the stream immediately, so a
    /// buffered read completes). RUNNING runs must use `runStream` instead — see Issue 3.
    func runStreamText(_ base: String, _ aid: String, _ runId: String) async throws -> String {
        let (data, _) = try await raw(base, "/api/agents/\(aid)/runs/\(runId)/stream")
        return String(decoding: data, as: UTF8.self)
    }

    /// Issue 3: incremental SSE reader for a RUNNING run's log. Yields one event per parsed
    /// `data:` frame; heartbeat comment lines (`:`) and blanks are skipped. The stream ends on
    /// the terminal `done` frame, an error, or cancellation (which cancels the URLSession task).
    func runStream(_ base: String, _ aid: String, _ runId: String) -> AsyncThrowingStream<RunStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let work = Task {
                do {
                    var request = try makeRequest(base, "/api/agents/\(aid)/runs/\(runId)/stream")
                    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    let (bytes, response) = try await streamSession.bytes(for: request)
                    guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
                    if Self.perimeterIntercepted(status: http.statusCode, contentType: http.value(forHTTPHeaderField: "Content-Type"), body: Data()) {
                        throw OrchaAuthRequiredError()
                    }
                    guard (200..<300).contains(http.statusCode) else {
                        throw OrchaApiError(status: http.statusCode, body: "")
                    }
                    for try await line in bytes.lines {
                        if Task.isCancelled { break }
                        if let event = Self.parseSseEvent(line) { continuation.yield(event) }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in work.cancel() }
        }
    }

    /// Parse a single SSE line into an event. Only `data:`-prefixed JSON frames yield an event.
    static func parseSseEvent(_ line: String) -> RunStreamEvent? {
        guard line.hasPrefix("data:") else { return nil }
        let payload = line.dropFirst(5).trimmingCharacters(in: .whitespaces)
        guard
            let data = payload.data(using: .utf8),
            let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        let seq = (obj["seq"] as? Int) ?? (obj["seq"] as? NSNumber)?.intValue ?? -1
        if (obj["done"] as? Bool) == true {
            return .done(seq: seq, status: (obj["status"] as? String) ?? "finished")
        }
        if let text = obj["line"] as? String { return .line(seq: seq, text: text) }
        return nil
    }

    // MARK: writes (actor ids per the human-authority contract)

    func postTaskMessage(_ base: String, _ tid: String, actor: String, body: String) async throws {
        try await post(base, "/api/tasks/\(tid)/messages", ["author_agent_id": actor, "body": body])
    }

    func cancelTask(_ base: String, _ tid: String, actor: String, reason: String?) async throws {
        try await post(base, "/api/tasks/\(tid)/cancel", ["actor_agent_id": actor, "reason": reason])
    }

    func verifyTask(_ base: String, _ tid: String, actor: String, approve: Bool, feedback: String?) async throws {
        try await post(base, "/api/tasks/\(tid)/verify", ["approve": approve, "feedback": feedback, "actor_agent_id": actor])
    }

    /// Collab v1 — name the human who should verify this task (nil = anyone; the
    /// absent key clears it, matching the server's Optional default). Owner-or-
    /// `assign_reviewers` gated server-side; advisory — /verify stays permissive.
    func setTaskReviewer(_ base: String, _ tid: String, actor: String, reviewerAgentId: String?) async throws {
        _ = try await send(base, "/api/tasks/\(tid)/reviewer", method: "PUT", [
            "actor_agent_id": actor,
            "reviewer_agent_id": reviewerAgentId,
        ])
    }

    func decidePlan(_ base: String, _ tid: String, actor: String, approve: Bool, reason: String?, target: String?) async throws {
        try await post(base, "/api/decisions", [
            "subject_type": "plan_approval",
            "subject_id": tid,
            "decision": approve ? "approve" : "reject",
            "reason": reason,
            "actor_agent_id": actor,
            "target_agent_id": target,
        ])
    }

    func respondRequest(_ base: String, _ rid: String, actor: String, response: String) async throws {
        try await post(base, "/api/requests/\(rid)/respond", ["responder_agent_id": actor, "response": response])
    }

    func closeRequest(_ base: String, _ rid: String, actor: String, reason: String?) async throws {
        try await post(base, "/api/requests/\(rid)/close", ["requester_agent_id": actor, "reason": reason])
    }

    /// GH #148 — the notifier (kill-switch). Human-only on the server; `actor` is the
    /// paired human, not an agent id.
    func setWakes(_ base: String, _ cid: String, actor: String, enabled: Bool) async throws {
        try await post(base, "/api/containers/\(cid)/wakes", ["enabled": enabled, "actor_agent_id": actor])
    }

    /// GH #148 — the autonomy gearbox (`plan` | `pr` | `full`). Human-gated on the server.
    func setAutonomy(_ base: String, _ cid: String, actor: String, level: String) async throws {
        try await post(base, "/api/containers/\(cid)/autonomy", ["level": level, "actor_agent_id": actor])
    }

    /// Bind (`repo` = "owner/name") or unbind (`repo` = nil) the container's GitHub repo.
    /// A nil repo makes `send` drop the key entirely — the server's schema defaults the
    /// absent field to null, so `{}` and `{"repo": null}` both unbind (github_routes.py).
    func setGithubRepo(_ base: String, _ cid: String, repo: String?) async throws -> GithubBindingResponse {
        try await putDecoding(base, "/api/containers/\(cid)/github", ["repo": repo])
    }

    /// Flow 07a: returns the routed outcome so the UI can tell a real wake
    /// (`nudged:true` → "Nudged {alias}") from the human-owns-it no-op (`nudged:false`).
    func nudgeRequest(_ base: String, _ rid: String, actor: String, note: String?) async throws -> NudgeResult {
        try await postDecoding(base, "/api/requests/\(rid)/nudge", ["actor_agent_id": actor, "note": note])
    }

    func escalateRequest(_ base: String, _ rid: String, actor: String, reason: String?) async throws {
        try await post(base, "/api/requests/\(rid)/escalate", ["requester_agent_id": actor, "reason": reason])
    }

    func acceptTaskRequest(_ base: String, _ rid: String, actor: String, note: String?) async throws {
        try await post(base, "/api/requests/\(rid)/accept-task", ["responder_agent_id": actor, "note": note])
    }

    func rejectTaskRequest(_ base: String, _ rid: String, actor: String, reason: String) async throws {
        try await post(base, "/api/requests/\(rid)/reject-task", ["responder_agent_id": actor, "reason": reason])
    }

    func convertRequest(_ base: String, _ rid: String, actor: String, title: String, dod: String, assignee: String?) async throws {
        try await post(base, "/api/requests/\(rid)/convert-to-task", [
            "requester_agent_id": actor,
            "title": title,
            "definition_of_done": dod,
            "assignee_alias": assignee,
        ])
    }


    func updateAgentModel(_ base: String, _ aid: String, model: String) async throws {
        try await post(base, "/api/agents/\(aid)/model", ["model": model])
    }

    func updateAutoWake(_ base: String, _ aid: String, actor: String, intervalSecs: Int?) async throws {
        try await patch(base, "/api/agents/\(aid)/auto-wake", ["actor_agent_id": actor, "interval_secs": intervalSecs])
    }

    func renameAgent(_ base: String, _ aid: String, actor: String, alias: String) async throws {
        try await patch(base, "/api/agents/\(aid)", ["actor_agent_id": actor, "alias": alias])
    }

    func retireAgent(_ base: String, _ aid: String, actor: String) async throws {
        try await post(base, "/api/agents/\(aid)/retire", ["actor_agent_id": actor])
    }

    func startConversation(_ base: String, _ aid: String, actor: String) async throws -> ConversationResponse {
        try await postDecoding(base, "/api/agents/\(aid)/conversations", ["actor_agent_id": actor])
    }

    func sendTurn(_ base: String, _ conversationId: String, actor: String, content: String) async throws {
        try await post(base, "/api/conversations/\(conversationId)/turns", [
            "role": "human", "author_agent_id": actor, "content": content,
        ])
    }

    func endConversation(_ base: String, _ conversationId: String, actor: String) async throws {
        try await post(base, "/api/conversations/\(conversationId)/end", ["actor_agent_id": actor])
    }

    func stopRun(_ base: String, _ runId: String, actor: String) async throws {
        try await post(base, "/api/runs/\(runId)/stop", ["actor_agent_id": actor])
    }

    func createTask(
        _ base: String, _ cid: String, actor: String,
        title: String, description: String?, dod: String,
        assignee: String?, priority: Int, dependsOn: [String], notReady: Bool
    ) async throws -> GenericIdResponse {
        try await postDecoding(base, "/api/containers/\(cid)/tasks", [
            "title": title,
            "description": description,
            "definition_of_done": dod,
            "priority": priority,
            "created_by_agent_id": actor,
            "assignee_alias": assignee,
            "depends_on": dependsOn,
            "not_ready": notReady,
        ])
    }

    // MARK: plumbing

    /// Every request is built here so the bearer credential (cloud auth perimeter's
    /// iOS/API lane) rides on ALL calls — reads, writes, and streams alike — whenever
    /// the target base URL has a token.
    private func makeRequest(_ base: String, _ path: String) throws -> URLRequest {
        guard let url = URL(string: base + path) else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        if let token = bearerOverride ?? BearerTokens.token(for: base) {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    /// The auth perimeter never answers portal JSON when the bearer credential is
    /// missing or wrong — the request falls through to the browser OAuth lane and
    /// comes back as a 401, or (after URLSession follows the sign-in redirects) an
    /// HTML page. Surface that as "needs an access token", never as a decode error.
    /// Portal-level errors (403 authority, 404, 409, 422 — all JSON) pass through.
    static func perimeterIntercepted(status: Int, contentType: String?, body: Data) -> Bool {
        if status == 401 { return true }
        if contentType?.lowercased().contains("text/html") == true { return true }
        let head = String(decoding: body.prefix(64), as: UTF8.self)
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        return head.hasPrefix("<!doctype html") || head.hasPrefix("<html")
    }

    private static func checkPerimeter(_ http: HTTPURLResponse, _ data: Data) throws {
        if perimeterIntercepted(
            status: http.statusCode,
            contentType: http.value(forHTTPHeaderField: "Content-Type"),
            body: data
        ) {
            throw OrchaAuthRequiredError()
        }
    }

    /// Build a `?a=1&b=2` suffix, dropping nil values and percent-encoding each value
    /// (ISO cursors carry `:`/`+` — over-encode to a conservative unreserved set).
    /// `internal` (not `private`): the per-feature `OrchaApiClient+*` extensions
    /// (github hub pagination/filter params) build query strings too.
    func query(_ items: [String: String?]) -> String {
        let pairs = items
            .sorted { $0.key < $1.key }
            .compactMap { key, value -> String? in
                guard let value else { return nil }
                let encoded = value.addingPercentEncoding(withAllowedCharacters: .orchaQueryValue) ?? value
                return "\(key)=\(encoded)"
            }
        return pairs.isEmpty ? "" : "?" + pairs.joined(separator: "&")
    }

    private func raw(_ base: String, _ path: String) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: makeRequest(base, path))
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        try Self.checkPerimeter(http, data)
        guard (200..<300).contains(http.statusCode) else {
            throw OrchaApiError(status: http.statusCode, body: String(decoding: data.prefix(300), as: UTF8.self))
        }
        return (data, http)
    }

    // `internal` (not `private`): the per-feature `OrchaApiClient+*` extensions live in
    // their own files and reach these two plumbing helpers to add endpoints (github hub).
    func get<T: Decodable>(_ base: String, _ path: String) async throws -> T {
        let (data, _) = try await raw(base, path)
        return try decoder.decode(T.self, from: data)
    }

    private func send(_ base: String, _ path: String, method: String, _ body: [String: Any?]) async throws -> Data {
        var request = try makeRequest(base, path)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let cleaned = body.compactMapValues { $0 }
        request.httpBody = try JSONSerialization.data(withJSONObject: cleaned)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        try Self.checkPerimeter(http, data)
        guard (200..<300).contains(http.statusCode) else {
            throw OrchaApiError(status: http.statusCode, body: String(decoding: data.prefix(300), as: UTF8.self))
        }
        return data
    }

    private func post(_ base: String, _ path: String, _ body: [String: Any?]) async throws {
        _ = try await send(base, path, method: "POST", body)
    }

    func postDecoding<T: Decodable>(_ base: String, _ path: String, _ body: [String: Any?]) async throws -> T {
        let data = try await send(base, path, method: "POST", body)
        return try decoder.decode(T.self, from: data)
    }

    private func patch(_ base: String, _ path: String, _ body: [String: Any?]) async throws {
        _ = try await send(base, path, method: "PATCH", body)
    }

    private func putDecoding<T: Decodable>(_ base: String, _ path: String, _ body: [String: Any?]) async throws -> T {
        let data = try await send(base, path, method: "PUT", body)
        return try decoder.decode(T.self, from: data)
    }
}

private extension CharacterSet {
    /// Conservative unreserved set for query VALUES — encodes `:`, `+`, `&`, `=`, space, etc.
    static let orchaQueryValue = CharacterSet(charactersIn:
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
}

/// The auth perimeter intercepted the request (bearer missing or rejected): the
/// caller needs a — correct — access token, not a different address.
struct OrchaAuthRequiredError: LocalizedError, Equatable {
    var errorDescription: String? {
        "This Orcha requires a team access token. Add or update it under Settings → Containers."
    }
}

struct OrchaApiError: LocalizedError {
    let status: Int
    let body: String

    var errorDescription: String? {
        switch status {
        case 403: "You don't have access to do this in this project."
        case 409: "Orcha rejected this action because the item changed. Refresh and try again."
        case 422: "Orcha needs more information for this action."
        default: "Orcha answered with an error (\(status))."
        }
    }
}
